[WIP] Some work on eager op binding codegen (gen_python_functions.py)

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32574 Fix/simplify alias annotation handling in op codegen.
* **#29986 [WIP] Some work on eager op binding codegen (gen_python_functions.py)**

Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

Commit:
Some work on eager op binding codegen (gen_python_functions.py)

Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: https://github.com/bhosmer/scratch/commit/1f2f07fb605e782cf7fdfb7d5eb33050eb65a6b4

Note: in the interests of keeping the diff contained, there's only
some light cleanup here beyond what's necessary for the codegen changes.
Plan is to do some more substantial refactoring in followup PRs that
leave generated code unchanged.

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: https://github.com/bhosmer/scratch/commit/1f2f07fb605e782cf7fdfb7d5eb33050eb65a6b4

Note: in the interests of keeping the diff contained, there's only
some light cleanup here beyond what's necessary for the codegen changes.
Plan is to do some more substantial refactoring in followup PRs that
leave generated code unchanged.

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: https://github.com/bhosmer/scratch/commit/1f2f07fb605e782cf7fdfb7d5eb33050eb65a6b4

Note: in the interests of keeping the diff contained, there's only
some light cleanup here beyond what's necessary for the codegen changes.
Plan is to do some more substantial refactoring in followup PRs that
leave generated code unchanged.

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

bhosmer commented:
cc @izdeby, LMK if there's anything I can do to stay out of the way of your upcoming TensorOptions stuff
cc @eellison FYI

zou3519 commented:
This might be a matter of personal preference, but I feel like the [older code is easier to read](https://github.com/bhosmer/scratch/commit/dc1416017668e9e97c6fd8efc13719f7c3db76aa#diff-b42b9646ea2d4f0005b3bece808baaeeR2022-R2041)

bhosmer commented:
> This might be a matter of personal preference, but I feel like the [older code is easier to read](https://github.com/bhosmer/scratch/commit/dc1416017668e9e97c6fd8efc13719f7c3db76aa#diff-b42b9646ea2d4f0005b3bece808baaeeR2022-R2041)

@zou3519 Hi Richard - I agree that the lambdas aren't ideal, if that's the main thing that's hitting you. They weren't in my original idea for bringing the code inline, but it turns out that the GIL needs to be reacquired before wrapping. 

Personally it felt like a net benefit to have the code all in one place, and the python is a bit simpler due to not having to coordinate a second positional param list (for the helpers). But I don't feel that strongly about it - the immediate focus is definitely more on taming the python than on tweaking the generated code. This change seemed incrementally better to me but I'm happy to keep generating the old style CG if folks prefer it.

zou3519 commented:
> @zou3519 Hi Richard - I agree that the lambdas aren't ideal, if that's the main thing that's hitting you. They weren't in my original idea for bringing the code inline, but it turns out that the GIL needs to be reacquired before wrapping.

My problem isn't really the lambdas, it's just that the new generated code seems harder to read because it doesn't have the helper functions. Take 
![image](https://user-images.githubusercontent.com/5652049/69076305-32198600-0a01-11ea-846d-9c900ea8af0a.png)
for instance. It's clear from a glance that we're asking the PythonArgParser to select an overload and then we are calling said overload. In the new code one needs to read a little more to see that that is what is happening.

Again, I think this might just be a matter of personal preference and I'd be happy to hear more opinions.

bhosmer commented:
> It's clear from a glance that we're asking the PythonArgParser to select an overload and then we are calling said overload. In the new code one needs to read a little more to see that that is what is happening.

Ah, gotcha. Agree on hearing more opinions, but regardless I think you're right that the overall select-and-call shape is a bit less prominent now. If I'm in here doing other stuff, I'll also see if generating a few well-placed comments helps restore some of that clarity.

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: https://github.com/bhosmer/scratch/commit/dc1416017668e9e97c6fd8efc13719f7c3db76aa

Note: in the interests of keeping the diff contained, there's only
some light cleanup here beyond what's necessary for the codegen changes.
Plan is to do some more substantial refactoring in followup PRs that
leave generated code unchanged.

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

izdeby commented:
> > It's clear from a glance that we're asking the PythonArgParser to select an overload and then we are calling said overload. In the new code one needs to read a little more to see that that is what is happening.
> 
> Ah, gotcha. Agree on hearing more opinions, but regardless I think you're right that the overall select-and-call shape is a bit less prominent now. If I'm in here doing other stuff, I'll also see if generating a few well-placed comments helps restore some of that clarity.

That's a tough call. On one hand i totally see Richards point of view, the code gets a bit more tricky but on the other hand we get a bunch of wins: 
- less generated files, less templates
- gen_python_functions.py  got a little cleaner 

Also, once you familiarize yourself with the new way of calling overloads it will get easier. 
So i would say its a step in the right direction.

izdeby commented:
@bhosmer, can you, please, kick off the fabricator build as well? just to see how all the jobs feel about this change

bhosmer commented:
> @bhosmer, can you, please, kick off the fabricator build as well? just to see how all the jobs feel about this change

I just updated D18567980 with the TARGETS changes to remove the `_dispatch.h` templates, so it's rebuilding/rerunning now... is that the build you mean? LMK if I'm completely unaware of some other part of the process :)

bhosmer commented:
@zou3519 FYI generates comments now, I think it helps a bit - thanks for the suggestion! (ofc clearly it's still more to read than before)

![image](https://user-images.githubusercontent.com/355313/69181919-e203f780-0ac4-11ea-8614-c0ffceaa1801.png)

resistor commented:
Does this have code size impact?

bhosmer commented:
> Does this have code size impact?

I doubt it but have been meaning to confirm - what should I look at, size of release-mode libtorch.so?

resistor commented:
> > Does this have code size impact?
> 
> @resistor I doubt it but have been meaning to confirm - what should I look at, size of release-mode libtorch.so?

Yeah, that's what I would check.

bhosmer commented:
Holding off on landing this pending some overlapping changes from @izdeby that should go in first. May add WIP commits in the meantime so no need to review yet. cc @gchanan

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/8f506d7293f22a9b1d69e33e86360a493a6bac92

Note: in the interests of keeping the diff contained, there's only
some light cleanup here beyond what's necessary for the codegen changes.
Plan is to do some more substantial refactoring in followup PRs that
leave generated code unchanged.

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/979d01d2eda4c104f56b50ed05f54a5e6b2bea07

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/979d01d2eda4c104f56b50ed05f54a5e6b2bea07

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/352f119bd0c913d89d02d47a6089495bea646bdc

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/352f119bd0c913d89d02d47a6089495bea646bdc

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 208f9c6:
* **1/1** failures introduced in this PR
* **1/1** recognized as flaky
    * Re-run these jobs?



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=208f9c6fbe9e521dea90d38663c64805e3b0dcc8).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4015584) pytorch_macos_10_13_cuda9_2_cudnn7_py3_build (1/1)
**Step:** "Update Homebrew" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77399945) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77399945))
<details>
<summary>
<code>fatal: The remote end hung up unexpectedly</code>
</summary>

```
++ [[ -z /usr/local ]] 
++ dir=/usr 
++ [[ -n '' ]] 
++ [[ -z /usr ]] 
++ dir= 
++ [[ -n '' ]] 
++ [[ -z '' ]] 
++ [[ -n '' ]] 
+ git fetch --depth=1 origin 
packet_write_wait: Connection to 140.82.113.4 port 22: Broken pipe  
fatal: The remote end hung up unexpectedly 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

gchanan commented:
are you updating the codegen diff?

bhosmer commented:
> are you updating the codegen diff?

Most definitely. Still rolling should be reviewable within a day or two, sorry for the delay :) might also put up a WIP in the meantime if it's in an informative state

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"

[WIP, day or two away from reviewability cc @gchanan]


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/commit/352f119bd0c913d89d02d47a6089495bea646bdc

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

gchanan commented:
Took a quick look at the generated code.  I really like some of the changes you made, e.g. the addition of the schemas, putting the dispatch code inline.

A couple of things:
1) I don't think changing some of the structure is safe.  For example, this diff:
```
+// _cufft_clear_plan_cache
 static PyObject * THPVariable__cufft_clear_plan_cache(PyObject* self_, PyObject* args, PyObject* kwargs)
 {
   HANDLE_TH_ERRORS
+
+
   static PythonArgParser parser({
     "_cufft_clear_plan_cache(int64_t device_index)",
   }, /*traceable=*/false);
-
   ParsedArgs<1> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);

-  if (r.idx == 0) {
-    dispatch__cufft_clear_plan_cache(r.toInt64(0));
-    Py_RETURN_NONE;
-  }
+  // aten::_cufft_clear_plan_cache(int device_index) -> ()
+
+  AutoNoGIL no_gil;
+  at::_cufft_clear_plan_cache(r.toInt64(0));
   Py_RETURN_NONE;
+
   END_HANDLE_TH_ERRORS
 }
```
you are dropping the GIL but not re-acquiring it before returning Py_NONE.

I assume changing some of the if/else structure is safe (like, python arg parser complains if none of the schemas match, right?).  For example:
```
   static PythonArgParser parser({
     "_cudnn_rnn_flatten_weight(TensorList weight_arr, int64_t weight_stride0, int64_t input_size, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first, bool bidirectional)",
   }, /*traceable=*/true);
-
   ParsedArgs<8> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);

-  if (r.idx == 0) {
-    return wrap(dispatch__cudnn_rnn_flatten_weight(r.tensorlist(0), r.toInt64(1), r.toInt64(2), r.toInt64(3), r.toInt64(4), r.toInt64(5), r.toBool(6), r.toBool(7)));
-  }
-  Py_RETURN_NONE;
+  // aten::_cudnn_rnn_flatten_weight(Tensor[] weight_arr, int weight_stride0, int input_size, int mode, int hidden_size, int num_layers, bool batch_first, bool bidirectional) -> Tensor
+
+  auto dispatch = [](TensorList weight_arr, int64_t weight_stride0, int64_t input_size, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first, bool bidirectional) {
+    AutoNoGIL no_gil;
+    return at::_cudnn_rnn_flatten_weight(weight_arr, weight_stride0, input_size, mode, hidden_size, num_layers, batch_first, bidirectional);
+  };
+
+  return wrap(dispatch(r.tensorlist(0), r.toInt64(1), r.toInt64(2), r.toInt64(3), r.toInt64(4), r.toInt64(5), r.toBool(6), r.toBool(7)));
+
   END_HANDLE_TH_ERRORS
 }
```

2) Can you fix the unnecessary spaces?  Makes it annoying to view at least in my version of git diff, which highlights it.

gchanan commented:
3) Naming the lambda dispatch -- is that safe if we add an operator argument named dispatch?

gchanan commented:
dispatch__index_copy_ takes a `const Tensor&` but the lambda takes a `Tensor` for self.  Was this an intentional change?

```
+// _index_copy_
 static PyObject * THPVariable__index_copy_(PyObject* self_, PyObject* args, PyObject* kwargs)
 {
   HANDLE_TH_ERRORS
+
+
   static PythonArgParser parser({
     "_index_copy_(Tensor input, int64_t dim, Tensor index, Tensor source)",
   }, /*traceable=*/true);
-
   ParsedArgs<4> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);

-  if (r.idx == 0) {
-    return wrap(dispatch__index_copy_(r.tensor(0), r.toInt64(1), r.tensor(2), r.tensor(3)));
-  }
-  Py_RETURN_NONE;
+  // aten::_index_copy_(Tensor(a!) self, int dim, Tensor index, Tensor source) -> Tensor(a!)
+
+  auto dispatch = [](Tensor self, int64_t dim, const Tensor & index, const Tensor & source) {
+    AutoNoGIL no_gil;
+    return at::_index_copy_(self, dim, index, source);
+  };
+
+  return wrap(dispatch(r.tensor(0), r.toInt64(1), r.tensor(2), r.tensor(3)));
+
   END_HANDLE_TH_ERRORS
 }
```

gchanan commented:
Interesting, here you fix the order of `out`:
```
-  if (r.idx == 0) {
-    if (r.isNone(1)) {
-      return wrap(dispatch_acos(r.tensor(0)));
-    } else {
-      return wrap(dispatch_acos(r.tensor(0), r.tensor(1)));
-    }
+  if (r.isNone(1)) {
+    // aten::acos(Tensor self) -> Tensor
+
+    auto dispatch = [](const Tensor & self) {
+      AutoNoGIL no_gil;
+      return self.acos();
+    };
+
+    return wrap(dispatch(r.tensor(0)));
   }
-  Py_RETURN_NONE;
+  else {
+    // aten::acos.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+
+    auto dispatch = [](Tensor out, const Tensor & self) {
+      AutoNoGIL no_gil;
+      return at::acos_out(out, self);
+    };
+
+    return wrap(dispatch(r.tensor(1), r.tensor(0)));
+  }
+
```

but here you don't:
```
+      else {
+        // aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
+
+        auto dispatch = [](const Tensor & self, Scalar alpha, const Tensor & other, Tensor out) {
+          AutoNoGIL no_gil;
+          return at::add_out(out, self, other, alpha);
+        };
+
+        return wrap(dispatch(r.tensor(0), r.scalar(1), r.tensor(2), r.tensor(3)));
+      }
     }
```

bhosmer commented:
> Took a quick look at the generated code. I really like some of the changes you made, e.g. the addition of the schemas, putting the dispatch code inline.

Thanks for taking a look! Unlike the python, the generated code shouldn't change much so this is time well spent.

> 
> A couple of things:
> 
> you are dropping the GIL but not re-acquiring it before returning Py_NONE.
>

Aha, thanks for catching this! Overeager on my part - I've never seen it blow up on tests but you're totally right, will restore the RAII here.

> 
> I assume changing some of the if/else structure is safe (like, python arg parser complains if none of the schemas match, right?). For example:
> 

Yeah - the old code lacked a catch-all else as well, so this should be a straight-across change, no new reliance on the parser's strictness.

> 
> 1. Can you fix the unnecessary spaces?  Makes it annoying to view at least in my version of git diff, which highlights it.

Sure thing :)

bhosmer commented:
> 1. Naming the lambda dispatch -- is that safe if we add an operator argument named dispatch?

Good point, probably not. I ran into the same issue with the `r` in an earlier version of the codegen that did more unpacking to locals - will add underscores to all the burned-in names (I'd rather do that than mangle)

bhosmer commented:
> dispatch__index_copy_ takes a `const Tensor&` but the lambda takes a `Tensor` for self. Was this an intentional change?
> 

Yeah, but I'll double check, it was an early change. Basically IIRC the change is because the current codegen doesn't observe the usual conversion of `Tensor &` to `Tensor` for `self`, but I'll track down why I think that is and give you a link to the old code.

bhosmer commented:
> Interesting, here you fix the order of `out`:
> 
> ```
> +    // aten::acos.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
> ```
> 
> but here you don't:
> 
> ```
> +        // aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
> ```

Oh man this one had me stumped - turns out it's the `call_args` override from deprecated.yaml. Adding a "(deprecated)" to the schema comment... :)

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"

[WIP, day or two away from reviewability cc @gchanan]


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"

[WIP, day or two away from reviewability cc @gchanan]


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

bhosmer commented:
> > dispatch__index_copy_ takes a `const Tensor&` but the lambda takes a `Tensor` for self. Was this an intentional change?
> 
> Yeah, but I'll double check, it was an early change. Basically IIRC the change is because the current (i.e., before this PR) codegen doesn't observe the usual conversion of `Tensor &` to `Tensor` for `self`, but I'll track down why I think that is and give you a link to the old code.

Hmmm actually here's what I see when I roll off my changes
```
inline Tensor dispatch__index_copy_(Tensor self, int64_t dim, const Tensor & index, const Tensor & source) {
```
Might have changed since my baseline, I haven't rebased in a while. Will recheck after that.

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"

[WIP, day or two away from reviewability cc @gchanan]


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"

[WIP, day or two away from reviewability cc @gchanan]


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

bhosmer commented:
@gchanan fyi rebased and in reviewable shape - earlier comments should all be addressed, final being the thing about `dispatch_index_copy` which looks the same after rebasing, so we should dig into what you're seeing...

BTW apologies for the big diff - happy to explain/motivate anything that's not obvious, also interactively if that's better.

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -736,105 +711,343 @@ def emit_namedtuple_return_type_def(declaration, next_index):
+
+""")
+
+# python binding for singe-overload function/method
```
nit: single

bhosmer replied:
Feeling burned by your comment

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -915,7 +1128,56 @@ def is_smaller(d1, d2):
+#
+
+SCHEMA_TYPE_CONVERSION_HACKS = {
+    'Type': 'ScalarType',
```
I really doubt this ever does anything.

bhosmer replied:
It was firing, but only on instances of `Type` introduced elsewhere in the codegen :)

gchanan replied:
that exists? :(

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -926,92 +1188,298 @@ def get_python_signature(declaration, include_out):
+# python binding args
+#
 
+# TODO blowtorch
```
hehe

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -926,92 +1188,298 @@ def get_python_signature(declaration, include_out):
+    return any([is_output(arg) for arg in declaration['arguments']])
+
+
+def get_tensor_options(declaration):
```
any reason you didn't add type annotations to these?

bhosmer replied:
You mean like mypy-style? Didn't really consider it... I guess I probably wouldn't have thought to add them to some things but not the whole file. Or is that what you mean? Probably wouldn't want to stop and do everything but happy to do a subset if you want (I've seen them elsewhere in the binding CG but if there's a convention for when to do it, I missed it)

gchanan replied:
Yes, I meant mypy-style.  I wasn't actually suggesting it, just curious if you thought about it.  There are some examples in function_wrapper.py and I find it makes it easier to read/reason about.

gchanan reviewed (approved):
looks great!

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -293,436 +158,546 @@ def get_py_torch_functions(declarations):
+#
+
+TEMP_SAFE_CPP_DECL_TYPE = {
+    'Tensor': 'const Tensor &',
```
I didn't look into this much, but this looks a little strange -- why does "Tensor" need a decl type transformation to be temp safe?

bhosmer replied:
I think I took the name (the conversion was already there) from the comment elsewhere in the file, about the dangling refs problem (Search for "Not good" :) But it didn't appear to be doing anything, so removing!

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -293,436 +158,546 @@ def get_py_torch_functions(declarations):
+    #
+    # You want:
+    #
+    #   Tensor dispatch_selu_(Tensor self) { return at::selu_(self); }
```
are you sure your dispatch lambda follows this pattern?  I just ask because the generated code doesn't use the explicit return syntax, so are you sure it's never inferred to be "Tensor&"?

gchanan replied:
at a minimum, you should probably update the comment to explain the situation after your changes.

bhosmer replied:
Thanks a bunch for catching this.
- added explicit return type for the lambdas - absolutely should've done this before, zero reason to omit it. Just an oversight. 
- updated the comment, again totally should have done this.

bhosmer commented on tools/autograd/gen_python_functions.py:
```diff
@@ -70,124 +45,160 @@
 if (r.isNone(${out_idx})) {
   ${call_dispatch}
-} else {
+}
```
Ahhh thanks for catching this. The odd formatting was a relic of an earlier version of the logic that split the if in a way that made this formatting simpler, but it's no longer necessary.

bhosmer commented on tools/autograd/gen_python_functions.py:
```diff
@@ -926,92 +1188,298 @@ def get_python_signature(declaration, include_out):
+    if has_tensor_options(declaration) or op_name(declaration).endswith('_like'):
+        return 'torch'
+    else:
+        return 'at'
```
Sure, added a comment. Tried `torch` everywhere just to check the easy change but there are definitely lots of ops that are `at` and not `torch` - but maybe you mean there might no longer be any reason not to move them over... FWIW here's the first 20 before the compiler gives up

```
no member named '_adaptive_avg_pool2d' in namespace 'torch'; did you mean 'at::_adaptive_avg_pool2d'?
no member named '_addr' in namespace 'torch'; did you mean 'at::_addr'?
no member named '_addr_out' in namespace 'torch'; did you mean 'at::_addr_out'?
no member named '_addr_' in namespace 'torch'; did you mean 'at::_addr_'?
no member named '_baddbmm_mkl_' in namespace 'torch'; did you mean 'at::_baddbmm_mkl_'?
no member named '_batch_norm_impl_index' in namespace 'torch'; did you mean 'at::_batch_norm_impl_index'?
no member named '_cast_Byte' in namespace 'torch'; did you mean 'at::_cast_Byte'?
no member named '_cast_Char' in namespace 'torch'; did you mean 'at::_cast_Char'?
no member named '_cast_Double' in namespace 'torch'; did you mean 'at::_cast_Double'?
no member named '_cast_Float' in namespace 'torch'; did you mean 'at::_cast_Float'?
no member named '_cast_Half' in namespace 'torch'; did you mean 'at::_cast_Half'?
no member named '_cast_Int' in namespace 'torch'; did you mean 'at::_cast_Int'?
no member named '_cast_Long' in namespace 'torch'; did you mean 'at::_cast_Long'?
no member named '_cast_Short' in namespace 'torch'; did you mean 'at::_cast_Short'?
no member named '_cat' in namespace 'torch'; did you mean 'at::_cat'?
no member named '_cat_out' in namespace 'torch'; did you mean 'at::_cat_out'?
no member named '_convolution' in namespace 'torch'; did you mean 'at::_convolution'?
no member named '_convolution_nogroup' in namespace 'torch'; did you mean 'at::_convolution_nogroup'?
no member named '_copy_from' in namespace 'torch'; did you mean 'at::_copy_from'?
```

gchanan replied:
Ah, I was thinking about this: https://github.com/pytorch/pytorch/blob/812b1ad8695c7bf28a8caa2f001f0ec3c554fbb7/torch/csrc/api/include/torch/types.h#L12-L31

I guess we don't include this anywhere outside of the "C++ API" (for some definition of the C++ API) and I'm not sure if it would cause problems.

bhosmer commented on tools/autograd/gen_python_functions.py:
```diff
@@ -293,436 +158,546 @@ def get_py_torch_functions(declarations):
+# addition to output-variant handler in which tensor options params
+# (if present) are checked against properties of a tensor output param
+# TODO remove hardcoding, use unpack logic from emit_single_dispatch
+PY_VARIABLE_CHECK_OUT_TYPE_HACK = CodeTemplate("""\
```
Totally agree that it's a problem having this check only on the eager side ofc. But as to whether or not we need it at all, AIUI the current situation is that the out variants of the factory functions (empty, zeros, ones, randn, etc) do all take the optional scattered TensorOptions args. 

Those args aren't made explicit in native_functions.yaml (if I found out why this is, I don't remember), but they're added in `gen_python_functions. make_python_binding_args` - i.e. the out-variant factories are one of the several small groups of ops that have these args added even though they don't specify a TensorOptions-typed arg.

(I thought you might have meant that these weren't documented, but looks like they are, e.g. https://pytorch.org/docs/stable/torch.html#torch.randn)

gchanan replied:
Sorry for not making this clear.  I was talking specifically about the combination of `out` arguments and the TensorOptions.  You can derive the type of one from the other, so it's pointless to have them both if you get the dispatch correct (and we don't have a C++ overload for them).  Also numpy doesn't have them.

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]

Commit:
Update on "[WIP] Some work on eager op binding codegen (gen_python_functions.py)"


Previously in addition to generating a python binding for each op,
we would generate an almost-trivial helper for each overload.
This PR eliminates the helpers, simplifying codegen logic a bit and
reducing the source-level indirection by a step.
Perf should be unchanged.

Also did some refactoring to the generation logic, should be a bit easier to follow.

codegen diff: 
https://github.com/bhosmer/scratch/compare/gen_python_functions

Differential Revision: [D18567980](https://our.internmc.facebook.com/intern/diff/D18567980)

[ghstack-poisoned]
